BAU: Update README after changes to createIdentityHandler (#72)
alphagov/passport-verify: README.md

# passport-verify

[![Build Status](https://travis-ci.org/alphagov/passport-verify.svg?branch=master)](https://travis-ci.org/alphagov/passport-verify)
[![Known Vulnerabilities](https://snyk.io/test/github/alphagov/passport-verify/badge.svg)](https://snyk.io/test/github/alphagov/passport-verify)
[![Greenkeeper badge](https://badges.greenkeeper.io/alphagov/passport-verify.svg)](https://greenkeeper.io/)

> `passport-verify` is a node and [passport.js](http://passportjs.org/) client for the [Verify Service Provider (VSP)](https://github.com/alphagov/verify-service-provider).

Author: ```suggestion
`passport-verify` is a Node.js and [passport.js](http://passportjs.org/) client for the [Verify Service Provider (VSP)](https://github.com/alphagov/verify-service-provider).
```

Reviewer: I want to commit your change but I don't know how :-/

Reviewer: Adding it all now as part of a separate commit 👍

// correspond to a the user's active session.
function loadRequestId (request) {

  // The following is an example that retrieves the request ID from the aforementioned
  // express-session object.
  return request.session.requestId
},
// Service Entity Id
'http://your-service-entity-id'
// This is only required when your Verify Service Provider is set up to be multi tenanted.
// If it is provided, it is passed to the Verify Service Provider with each request, and
// used to identify this service.

// Saml Form Template Location
'saml-form-template.njk',
// This is an optional parameter which can be used to style the saml form
// used to send the authn request to Verify.
// This template should only be rendered if Javascript has been disabled in the user's browser.
// If provided, the ssoLocation and samlRequest recieved from the Verify Service Provider
// will be provided to the named template for rendering.
// If this is not provided, passport-verify will render a default auto posting form
// with the correct attributes.

// Level of Assurance
// LEVEL_1 or LEVEL2 depending on your service's requirements. Defaults to LEVEL_2.
'LEVEL_2'
))
```

> If your service connects to GOV.UK Verify using a Matching Service Adapter, you should use the `createStrategy` method to create a strategy when configuring passport.js.

Author: Services don't connect to Verify using MSA, but they they use it for matching users, so this migh be more accurate?
```suggestion
If your service connects to GOV.UK Verify and uses the [Verify Matching Service Adapter](https://github.com/alphagov/verify-matching-service-adapter/), you should use the `createStrategy` method to create a strategy when configuring passport.js.
```

Reviewer: Noted. Will apply change 👍

> See the [API documentation for the createStrategy method](https://alphagov.github.io/passport-verify/modules/_passport_verify_strategy_.html#createstrategy) for more details.

Author: I wonder if this should be part of the previous paragraph, because it only applies to the people with MSA.

Reviewer: Yes, that makes sense. I had it separate to try and improve readability because it seemed like a really long paragraph. But it's not that long when it's published. Only the markdown is miles long because of the link...

The exact routes depend on how you plan on using the responses from GOV.UK Verify Hub. The example in the code block configures a route to allow a verified user to access the service:

```javascript

// route for authenticating a user
app.post('/verify/start', passport.authenticate('verify'))

// route for handling a callback from verify
app.post('/verify/response', (req, res, next) => (

   // in this example, authenticate() is being called from within the route handler
   // rather than being used as middleware, this provides access to the request
   // and response objects through closure
   const authMiddleware = passport.authenticate('verify', function (error, identity, infoOrError, status) {

    if (error) {
      return res.send(`TODO: render error-page with message ${error: error.message}`)
    }

    if (identity) {
      // passport-verify requires the use of a custom callback to handle successful
      // authentication
      return req.logIn(identity, () => res.send('TODO: redirect to service landing page')))
    }

    return res.send(`TODO: redirect to authentication failed page with ${error: infoOrError}`)

  })
  authMiddleware(req, res, next)
```

> If your service connects to GOV.UK Verify using a Matching Service Adapter, you should use the `createResponseHandler` method to configure routes for the authentication flow.

Author: same as my comments above?

Reviewer: Yup 👍

# passport-verify

[![Build Status](https://travis-ci.org/alphagov/passport-verify.svg?branch=master)](https://travis-ci.org/alphagov/passport-verify)
[![Known Vulnerabilities](https://snyk.io/test/github/alphagov/passport-verify/badge.svg)](https://snyk.io/test/github/alphagov/passport-verify)
[![Greenkeeper badge](https://badges.greenkeeper.io/alphagov/passport-verify.svg)](https://greenkeeper.io/)

`passport-verify` is a node and [passport.js](http://passportjs.org/) client for the [Verify Service Provider (VSP)](https://github.com/alphagov/verify-service-provider).

## Before you start

[Set up the VSP](https://github.com/alphagov/verify-service-provider/README.md).

## Usage

### Step 1. Install`passport-verify`

```bash
npm install --save passport-verify
```

### Step 2. Configure the `passport-verify` strategy

> **TODO** To configure the strategy you need the info in the code blocks when following instructions on here => [Configure `passport-verify` strategy](http://passportjs.org/docs/configure#strategies).

Reviewer 2: Andrea, you were planning to return to this, right?

Reviewer: yup, I've addressed it in the subsequent commit 👌 

it was only later when I realised what the TODOs in the examples were about - it's where users need to fill in details!... This is not common practice in my (limited) experience, but happy for them to stay in if it's common in the Node.js community.

<details>
<summary>
See an example of how to call the `createStrategy` method
</summary>

```javascript
// route for authenticating a user
app.post('/verify/start', passport.authenticate('verify'))

// route for handling a callback from verify
app.post('/verify/response', (req, res, next) => (

 // in this example, authenticate() is being called from within the route handler
 // rather than being used as middleware, this provides access to the request
 // and response objects through closure
 const authMiddleware = passport.authenticate('verify', function (error, user, infoOrError, status) {

  if (error) {
    return res.send(`TODO: render error-page with message ${error: error.message}`)
  }

  if (user) {
    // passport-verify requires the use of a custom callback to handle successful
    // authentication
    return req.logIn(user, () => res.send('TODO: redirect to service landing page')))
  }

  return res.send(`TODO: redirect to authentication failed page with ${error: infoOrError}`)

})
authMiddleware(req, res, next)
```
</details>

> 

Reviewer 2: I think this section will read more clearly if you can add a line break or two after the folds.

Reviewer: I agree!
But @jakubmiarka  and I had decided to remove example code blocks for anything that's for the legacy setup, since they won't be primary users. Still keeping the links to the API reference for everything, which is comprehensive enough.

## Usage

> The information here applies to services connecting using VSP 2.0.0 or above. If your connection involves a Matching Service Adapter

Reviewer 2: The second sentence appears to be unfinished.  As Linguo would say - "sentence fragment".

Reviewer: Oh, I missed that one 🙈 Fixing now...

If you are using this package for your application, note that the `DEBUG` variable will be read as a comma seperated list, so you can add or remove passport-verify logs as necessary without changing your own.

## API

See [the API documentation](https://alphagov.github.io/passport-verify/modules/_passport_verify_.html) for more details.

## Terminology

 * _Identity Provider_ is a service that can authenticate users
 * _Relying party_ is a service that needs to authenticate users
 * [_Verify Service Provider_](https://github.com/alphagov/verify-service-provider)
    is a service that consumes and produces SAML messages that can be used to communicate with GOV.UK Verify

>  * [_Passport.js_](http://passportjs.org/) is a node js library that provides a generic authentication framework for various authentication providers.

Author: ```suggestion
 * [_Passport.js_](http://passportjs.org/) is a Node.js library that provides a generic authentication framework for various authentication providers.
```

Reviewer: Yup 👍 👍
